pkg/proxy/dns.go

```diff
@@ -4,46 +4,37 @@
 package proxy
 
 import (
+	"errors"
+
 	"github.com/sirupsen/logrus"
 
-	"github.com/cilium/cilium/pkg/fqdn/proxy"
-	"github.com/cilium/cilium/pkg/fqdn/restore"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/policy"
+	defaultdns "github.com/cilium/cilium/pkg/proxy/defaultdns"
 	"github.com/cilium/cilium/pkg/revert"
 )
 
-// DefaultDNSProxy is the global, shared, DNS Proxy singleton.
-var DefaultDNSProxy proxy.DNSProxier
-
 // dnsRedirect implements the Redirect interface for an l7 proxy
 type dnsRedirect struct {
 	Redirect
-	proxyRuleUpdater proxyRuleUpdater
+	dnsProxy *defaultdns.Proxy
 }
 
 func (dr *dnsRedirect) GetRedirect() *Redirect {
 	return &dr.Redirect
 }
 
-// proxyRuleUpdater updates L7 proxy rules per endpoint.
-//
-// Note: Implementations must not take the IPcache lock, as the usage of this
-// interface is within the endpoint regeneration critical section.
-type proxyRuleUpdater interface {
-	// UpdateAllowed updates the rules in the DNS proxy with newRules for the
-	// endpointID and destPort.
-	UpdateAllowed(endpointID uint64, destPort restore.PortProto, newRules policy.L7DataMap) (revert.RevertFunc, error)
-}
-
 // setRules replaces old l7 rules of a redirect with new ones.
 func (dr *dnsRedirect) setRules(newRules policy.L7DataMap) (revert.RevertFunc, error) {
 	dr.logger.Debug(
 		"DNS Proxy updating matchNames in allowed list during UpdateRules",
 		logfields.NewRules, newRules,
 		logfields.EndpointID, dr.endpointID,
 	)
-	return dr.proxyRuleUpdater.UpdateAllowed(uint64(dr.endpointID), dr.dstPortProto, newRules)
+	if dnsProxy := dr.dnsProxy.Get(); dnsProxy != nil {
+		return dnsProxy.UpdateAllowed(uint64(dr.endpointID), dr.dstPortProto, newRules)
```

Comment:
nit: go style generally prefers handling the error case in the indented branch of the if; https://go.dev/wiki/CodeReviewComments#indent-error-flow